tests/sentry/snuba/metrics/test_query.py

```diff
@@ -0,0 +1,326 @@
+            "start": self.start,
+            "end": self.end,
+            "granularity": self.granularity,
+            "select": self.select,
+            "orderby": self.orderby,
+            "where": self.where,
+            "groupby": self.groupby,
+            "limit": self.limit,
+            "offset": self.offset,
+            "histogram_buckets": self.histogram_buckets,
+            "include_series": self.include_series,
+            "include_totals": self.include_totals,
+        }
+
+
+def test_validate_select():
+    with pytest.raises(InvalidParams, match='Request is missing a "field"'):
+        MetricsQuery(**MetricsQueryBuilder().with_select([]).to_metrics_query_dict())
+
+    with pytest.raises(
+        InvalidParams,
+        match=(
+            "Invalid operation 'foo'. Must be one of avg, count_unique, count, max, sum, "
+            "histogram, p50, p75, p90, p95, p99"
+        ),
+    ):
+        MetricsQuery(
+            **MetricsQueryBuilder()
+            .with_select([MetricField(op="foo", metric_name=SessionMetricKey.DURATION.value)])
+            .to_metrics_query_dict()
+        )
+    with pytest.raises(
+        DerivedMetricParseException,
+        match=(
+            re.escape(
+                "Failed to parse sum(session.crash_free_rate). No operations can be "
+                "applied on this field as it is already a derived metric with an "
+                "aggregation applied to it."
+            )
+        ),
+    ):
+        MetricsQuery(
+            **MetricsQueryBuilder()
+            .with_select(
+                [MetricField(op="sum", metric_name=SessionMetricKey.CRASH_FREE_RATE.value)]
+            )
+            .to_metrics_query_dict()
+        )
+
+
+def test_validate_order_by():
+    metric_field_2 = MetricField(op=None, metric_name=SessionMetricKey.ALL.value)
+    metrics_query_dict = (
+        MetricsQueryBuilder()
+        .with_orderby(
+            [
+                OrderBy(field=MetricsQueryBuilder.AVG_DURATION_METRIC, direction=Direction.ASC),
+                OrderBy(field=metric_field_2, direction=Direction.ASC),
+            ]
+        )
+        .to_metrics_query_dict()
+    )
+    with pytest.raises(InvalidParams, match="Only one 'orderBy' is supported"):
+        MetricsQuery(**metrics_query_dict)
+
+    with pytest.raises(
+        InvalidParams,
+        match=(
+            "Invalid operation 'foo'. Must be one of avg, count_unique, count, max, sum, "
+            "histogram, p50, p75, p90, p95, p99"
+        ),
+    ):
+        MetricsQuery(
+            **MetricsQueryBuilder()
+            .with_orderby(
+                [
+                    OrderBy(
+                        field=MetricField(op="foo", metric_name=SessionMetricKey.DURATION.value),
+                        direction=Direction.ASC,
+                    )
+                ]
+            )
+            .to_metrics_query_dict()
+        )
+    with pytest.raises(
```

Comment:
let's split tests into smaller ones

Reply:
I don't think that's necessary. they are small enough tests in functionality that I would rather keep them like this

Reply:
problem with this approach, you will never check second assert.

Reply:
how so?